src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs

```diff
@@ -0,0 +1,279 @@
+        {
+            CodeActionResolveData data;
+            if (codeAction.Data is CodeActionResolveData codeActionResolveData)
+            {
+                data = codeActionResolveData;
+            }
+            else
+            {
+                data = ((JToken)codeAction.Data).ToObject<CodeActionResolveData>();
+            }
+
+            var document = SolutionProvider.GetDocument(data.CodeActionParams.TextDocument, clientName);
+            var codeActions = await CodeActionsHandler.GetCodeActionsAsync(
+                document,
+                _codeFixService,
+                _codeRefactoringService,
+                data.CodeActionParams.Range,
+                cancellationToken).ConfigureAwait(false);
+
+            if (codeActions == null || !codeActions.Any())
+            {
+                return codeAction;
+            }
+
+            var codeActionToResolve = GetCodeActionToResolve(data.DistinctTitle, codeActions);
+
+            // We didn't find a matching action, so just return the action without an edit or command.
+            if (codeActionToResolve == null)
+            {
+                return codeAction;
+            }
+
+            var operations = await codeActionToResolve.GetOperationsAsync(cancellationToken).ConfigureAwait(false);
+            if (operations.IsEmpty)
+            {
+                return codeAction;
+            }
+
+            // TO-DO:
+            // 1) We currently must execute code actions which add new documents on the server as commands,
+            // since there is no LSP support for adding documents yet. In the future, we should move these actions
+            // to execute on the client.
+            // 2) There is also a bug (same tracking item) where code actions that edit documents other than the
+            // one where the code action was invoked from do not work. We must temporarily execute these as commands
+            // as well.
+            // https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1147293/
+            var runAsCommand = false;
+
+            var applyChangesOperations = operations.Where(operation => operation is ApplyChangesOperation);
+            if (applyChangesOperations.Any())
+            {
+                var workspaceEdits = await ComputeWorkspaceEdits(applyChangesOperations, document!, cancellationToken).ConfigureAwait(false);
+                if (workspaceEdits.Any())
+                {
+                    codeAction.Edit = new LSP.WorkspaceEdit { DocumentChanges = workspaceEdits };
+                }
+                else
+                {
+                    // The workspace edit is something we don't currently support, like adding a new document.
+                    runAsCommand = true;
+                }
+            }
+
+            // Set up to run as command on the server instead of using WorkspaceEdits.
+            var commandOperations = operations.All(operation => !(operation is ApplyChangesOperation));
```

Comment:
Ah I see, we apply any non ApplyChangesOperation as a command.  Then I think I'm only worried about the order.  The LSP spec states that commands are applied after workspace edits.  But if the code action specifies the non-ApplyChangesOperation to run before the ApplyChangesOperation, the code action could get corrupted if order matters.

Reply:
This also could be problematic - I see in the execute code actions request, we just find the one with the right title.  It will end up executing all of the operations, so if there are workspace edits + a command, it will execute the edits twice (once on client workspace edits, once on server during RunCodeActions ApplyChangesOperation)

Reply:
Do you think it would be best if we were to just file a Roslyn bug on this?
I think ideally, the solution would be to include a check in RunCodeActionsHandler to see whether we are executing an ApplyChangesOperation so that we don't execute those kinds of operations twice. But we can't currently do that since some ApplyChangesOperations are applied as commands due to [this](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1147293/) bug on the LSP side. Once the LSP bug is fixed, I think we can address the concerns you mentioned.

The current approach is: if we have a mix of ApplyChangesOperations and other operations, only the ApplyChangesOperations are executed. But if we have only non-ApplyChangesOperations, then we can execute them (as a command). This isn't ideal, but it prevents workspace edits from potentially executing twice.